File: TypeScript/app/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality() {
    this.items.forEach(item => createSellableItemWith(item).updateQuality());
    return this.items;
  }
}

function createSellableItemWith(item: Item): SellableItem {

  const sellableItemNames = {
    'Sulfuras, Hand of Ragnaros': SulfurasItem,
    'Aged Brie': AgedBrieItem,
    'Backstage passes to a TAFKAL80ETC concert': BackstagePasseItem,
    'Conjured Mana Cake': ConjuredItem
  }

  return sellableItemNames[item.name] !== undefined ? new sellableItemNames[item.name](item) : new SellableItem(item);
}

class SellableItem {
  constructor(protected item: Item) {}

  updateQuality(): void {
    this.decreaseItemQuality();
    this.decreaseItemSellIn();

    if (this.itemDoesNotHaveSellIn()) {
      this.decreaseItemQuality();
    }
  }

  decreaseItemSellIn() {
    this.item.sellIn = this.item.sellIn - 1;
  }

  decreaseItemQuality() {
    const item = this.item;
    if (!itemHasQuality()) return;
    this.item.quality = this.item.quality - 1

    function itemHasQuality() {
      return item.quality > 0;
    }
  }

  itemDoesNotHaveSellIn() {
    return this.item.sellIn < 0;
  }

  increaseItemQuality() {
    const item = this.item;
    if (!itemQualityIsBelowTheLimit()) return;
    item.quality = item.quality + 1

    function itemQualityIsBelowTheLimit() {
      const qualityLimit = 50;
      return item.quality < qualityLimit;
    }
  }
}

class SulfurasItem extends SellableItem {

  updateQuality(): void {
    return;
  }
}

class AgedBrieItem extends SellableItem {

  updateQuality(): void {
    this.increaseItemQuality();
    this.decreaseItemSellIn();
    if (this.itemDoesNotHaveSellIn()) this.increaseItemQuality()
  }

}

class BackstagePasseItem extends SellableItem {

  updateQuality() {
    const item = this.item;
    this.increaseItemQuality();
    if (itemIsBelowFirstConcertProximityThreshold()) {
      this.increaseItemQuality();
    }
    if (itemIsBelowSecondConcertProximityThreshold()) {
      this.increaseItemQuality();
    }
    this.decreaseItemSellIn();

    if (this.itemDoesNotHaveSellIn()) {
      clearItemQuality();
    }

    function itemIsBelowFirstConcertProximityThreshold() {
      const firstConcertProximityThreshold = 11;
      return item.sellIn < firstConcertProximityThreshold;
    }

    function itemIsBelowSecondConcertProximityThreshold() {
      const secondConcertProximityThreshold = 6;
      return item.sellIn < secondConcertProximityThreshold;
    }

    function clearItemQuality() {
      item.quality = 0
    }
  }
}

class ConjuredItem extends SellableItem {

  updateQuality(): void {
    this.decreaseItemQuality();
    this.decreaseItemQuality();
    this.decreaseItemSellIn();
    if (this.itemDoesNotHaveSellIn()) {
      this.decreaseItemQuality();
    }
  }

}

File: csharpcore/GildedRose/GildedRose.cs
using System.Collections.Generic;

namespace GildedRoseKata
{
    public class GildedRose
    {
        readonly IList<Item> _items;
        public GildedRose(IList<Item> items)
        {
            _items = items;
        }

        public void UpdateQuality()
        {
            foreach (var item in _items)
            {
                CreateGildedRoseItemFrom(item).Update();
            }
        }

        private Regular CreateGildedRoseItemFrom(Item item)
        {
            return (item.Name) switch
            {
                "Aged Brie" => new AgedBrie(item),
                "Backstage passes to a TAFKAL80ETC concert" => new BackstagePass(item),
                "Sulfuras, Hand of Ragnaros" => new Sulfuras(item),
                "Conjured" => new Conjured(item),
                _ => new Regular(item)
            };
        }

        private class Conjured : Regular
        {
            public Conjured(Item item) : base(item)
            {
            }

            public override void Update()
            {
                DecreaseQuality();
                DecreaseQuality();
                DecreaseSellIn();
                if(HasSellDateBeenPassed())
                {
                    DecreaseQuality();
                    DecreaseQuality();
                }
            }
        }

        private class Regular
        {
            protected readonly Item Item;

            public Regular(Item item)
            {
                Item = item;
            }

            public virtual void Update()
            {
                DecreaseQuality();
                DecreaseSellIn();
                if (HasSellDateBeenPassed())
                {
                    DecreaseQuality();
                }
            }
            
            protected void DecreaseQuality()
            {
                if (Item.Quality > 0)
                {
                    Item.Quality -= 1;
                }
            }

            protected bool HasSellDateBeenPassed()
            {
                return Item.SellIn < 0;
            }

            protected void DecreaseSellIn()
            {
                Item.SellIn -= 1;
            }
            
            protected void IncreaseQuality()
            {
                if (Item.Quality < 50)
                {
                    Item.Quality += 1;
                }
            }
        }

        private class AgedBrie : Regular
        {
            
            public AgedBrie(Item item) : base(item)
            {
            }
            
            public override void Update()
            {
                IncreaseQuality();
                DecreaseSellIn();
                if (HasSellDateBeenPassed())
                {
                    IncreaseQuality();
                }
            }
        }

        class BackstagePass : Regular
        {

            public BackstagePass(Item item) : base(item)
            {
            }

            public override void Update()
            {
                IncreaseQuality();
                if (isTheConcertBelowTheFirstThreshold())
                {
                    IncreaseQuality();
                }
                if (isTheConcertBelowSecondThreshold())
                {
                    IncreaseQuality();
                }
                DecreaseSellIn();

                if (HasSellDateBeenPassed())
                {
                    clearItemQuality();
                }

                bool isTheConcertBelowTheFirstThreshold()
                {
                    return Item.SellIn < 11;
                }

                bool isTheConcertBelowSecondThreshold()
                {
                    return Item.SellIn < 6;
                }

                void clearItemQuality()
                {
                    Item.Quality = 0;
                }
            }
        }

        class Sulfuras : Regular
        {
            public Sulfuras(Item item) : base(item)
            {
            }
            
            public override void Update()
            {
            }
        }
    }
}
